src/v/crypto/ossl_context_service.cc

```diff
@@ -0,0 +1,288 @@
+    auto init_return = initialize_openssl(
+      ctx, module_path, conf_file, fips_mode);
+
+    if (init_return.has_failure()) {
+        return init_return.as_failure();
+    }
+
+    return initialize_thread_return{
+      .init_ret = std::move(init_return.assume_value()),
+      .orig_ctx = old_context};
+}
+
+void finalize_worker_thread(OSSL_LIB_CTX* orig_ctx) {
+    OSSL_LIB_CTX_set0_default(orig_ctx);
+}
+} // namespace
+
+ossl_context_service::~ossl_context_service() noexcept = default;
+
+class ossl_context_service::impl final {
+    friend class ossl_context_test_class;
+
+public:
+    impl(
+      ssx::singleton_thread_worker& thread_worker,
+      ss::sstring config_file,
+      ss::sstring module_path,
+      is_fips_mode fips_mode)
+      : _thread_worker(thread_worker)
+      , _config_file(std::move(config_file))
+      , _module_path(std::move(module_path))
+      , _fips_mode(fips_mode) {}
+
+    ~impl() noexcept {
+        vassert(
+          _shard_ctx == nullptr && _old_context == nullptr,
+          "OpenSSL context service being destructed without being properly "
+          "shutdown");
+    }
+
+    impl(const impl&) = delete;
+    impl& operator=(const impl&) = delete;
+    impl(impl&&) noexcept = delete;
+    impl& operator=(impl&&) noexcept = delete;
+
+    ss::future<> start() {
+        lg.trace("Starting service...");
+        if (ss::this_shard_id() == 0) {
+            // On main shard, load the 'null' provider to the default context
+            // This prevents the default context from performing any
+            // cryptographic operation
+            _defctxnull = internal::OSSL_PROVIDER_ptr(
+              OSSL_PROVIDER_load(nullptr, "null"));
+            lg.debug("Loaded null into global provider");
+            // We also need to create a library contxt and load it in just
+            // within the thread worker's thread.  This is so the krb5 library
+            // will use this loaded context for any operations it needs to
+            // perform
+            _thread_worker_ctx = internal::OSSL_LIB_CTX_ptr(OSSL_LIB_CTX_new());
+            auto init_resp = co_await _thread_worker.submit([this] {
+                return initialize_worker_thread(
+                  _thread_worker_ctx.get(),
+                  _module_path,
+                  _config_file,
+                  _fips_mode);
+            });
+
+            if (init_resp.has_error()) {
+                throw exception(init_resp.assume_error());
+            }
+
+            _initialize_thread_worker_holder = std::move(
+              init_resp.assume_value());
+        }
+
+        _shard_ctx = internal::OSSL_LIB_CTX_ptr(OSSL_LIB_CTX_new());
+        _old_context = OSSL_LIB_CTX_set0_default(_shard_ctx.get());
+        lg.debug(
+          "Created new shard context for {} replacing {}",
+          fmt::ptr(_shard_ctx.get()),
+          fmt::ptr(_old_context));
+        auto init_resp = co_await _thread_worker.submit([this] {
```

Comment:
question: How does this intended to work? The memory for the contexts is allocated on the correct shard but then the work to initialize and set them is dispatched to thread worker which may not run the function on the calling shard.

Reply:
Memory allocated in an alien thread uses the system allocator, not the seastar allocator. The seastar allocator has a slow path for freeing memory not allocated on that shard, but it's generally not recommended.

Reply:
+1 , i was referring to `_shard_ctx` which is allocated in with seastar shard, due to the call being a few lines above the call to `.submit`

Reply:
Allocation and assignment of the OpenSSL library context happens per shard.  The initialization of the library context needs to happen in an alien thread because there's a bunch of system call stuff that OpenSSL does under the hood (namely dlloading the providers and opening and reading files, etc).

Reply:
Chatted with @graphcareful , I'll add some comments that hopefully will clear this up :)